WrightTools/artists/_interact.py

```
@@ -0,0 +1,381 @@
+
+
+def norm(arr, signed, ignore_zero=True):
+    if signed:
+        norm = np.nanmax(np.abs(arr))
+    else:
+        norm = np.nanmax(arr)
+    if norm != 0 and ignore_zero:
+        arr /= norm
+    return arr
+
+
+def set_aspect(xaxis, yaxis):
+    if xaxis.units == yaxis.units:
+        xr = xaxis.max() - xaxis.min()
+        yr = yaxis.max() - yaxis.min()
+        aspect = np.abs(yr / xr)
+        if 3 < aspect or aspect < 1 / 3.:
+            raise Warning(
+                "units agree, but aspect {0} required for equal spacing is too"
+                / + "narrow.".format(aspect)
+            )
+            aspect = np.clip(aspect, 1 / 3., 3.)
+            print("using aspect {0} instead".format(aspect))
+    else:
+        aspect = 1
+    return aspect
+
+
+def interact2D(data, xaxis=0, yaxis=1, channel=0, local=False, verbose=True):
+    """ Interactive 2D plot of the dataset.
+    Side plots show x and y projections of the slice (shaded gray).
+    Left clicks on the main axes draw 1D slices on side plots at the coordinates selected.
+    Right clicks remove the 1D slices.
+    For 3+ dimensional data, sliders below the main axes are used to change which slice is viewed.
+
+    Parameters
+    ----------
+    data : WrightTools.Data object
+        Data to plot.
+    axis : string, integer, or data.Axis object (optional)
+        Expression or index of axis. Default is 0.
+    channel : string, integer, or data.Channel object (optional)
+        Name or index of channel to plot. Default is 0.
+    local : boolean (optional)
+        Toggle plotting locally. Default is False.
+    verbose : boolean (optional)
+        Toggle talkback. Default is True.
+    """
+    # unpack
+    channel = get_channel(data, channel)
+    xaxis, yaxis = get_axes(data, [xaxis, yaxis])
+    cmap = get_colormap(channel)
+    current_state = Bunch()
+    # create figure
+    nsliders = data.ndim - 2
+    if nsliders < 0:
+        print("note enough dimensions")
```

raise exception, rather than print and return `wt.exceptions.DimensionalityError` exists for this exact purpose